DigitalLearningSolutions.Data/DataServices/EmailVerificationDataService.cs

```diff
@@ -0,0 +1,76 @@
+﻿namespace DigitalLearningSolutions.Data.DataServices
+{
+    using System;
+    using System.Data;
+    using System.Linq;
+    using Dapper;
+
+    public interface IEmailVerificationDataService
+    {
+        int CreateEmailVerificationHash(string hash, DateTime created);
+
+        void UpdateEmailVerificationHashIdForPrimaryEmail(int userId, int hashId);
+
+        void UpdateEmailVerificationHashIdForCentreEmail(int userId, int centreId, int hashId);
+
+        bool AccountEmailIsVerifiedForUser(int userId, string? email);
+    }
+
+    public class EmailVerificationDataService : IEmailVerificationDataService
+    {
+        private readonly IDbConnection connection;
+
+        public EmailVerificationDataService(IDbConnection connection)
+        {
+            this.connection = connection;
+        }
+
+        public int CreateEmailVerificationHash(string hash, DateTime created)
+        {
+            return connection.QuerySingle<int>(
+                @"INSERT INTO EmailVerificationHashes (EmailVerificationHash, CreatedDate)
+                        OUTPUT INSERTED.ID
+                        VALUES (@hash, @created)",
+                new { hash, created }
+            );
+        }
+
+        public void UpdateEmailVerificationHashIdForPrimaryEmail(int userId, int hashId)
+        {
+            connection.Execute(
+                @"UPDATE Users SET EmailVerificationHashID = @hashId WHERE ID = @userId",
+                new { hashId, userId }
+            );
+        }
+
+        public void UpdateEmailVerificationHashIdForCentreEmail(int userId, int centreId, int hashId)

```

Comment:
I recommend you change this to
```
        public void UpdateEmailVerificationHashIdForCentreEmails(int userId, string emailAddress, int hashId)
        {
            connection.Execute(
                @"UPDATE UserCentreDetails SET EmailVerificationHashID = @hashId WHERE UserID = @userId AND Email = @emailAddress",
                new { hashId, userId, emailAddress }
            );
        }
```
to ensure that the EmailVerificationHashID is only set when the email address is what we expect it to be, and to allow us to update all matching `UserCentreDetails` records at once
```